#  Pandas : Python Library

## What is pandas ?

##### Pandas is a Python library used for data analysis and manipulation. It provides data structures like DataFrames (tables) and Series (one-dimensional arrays) that allow efficient storage and processing of structured data.
##### Pandas offers tools for tasks such as data cleaning, transformation, merging, and aggregation, making it a fundamental library for data scientists and analysts working with Python. 

## Importing Pandas

In [1]:
import pandas as pd

### Loading Files 

In [2]:
df = pd.read_csv("C:\\Users\\Administrator\\Downloads\\pandas_practice1_dataset.csv")

## Functions 

#### 1. head() : display first five rows.

In [3]:
df.head()

Product_ID Product_Name     Brand     Category    Price  Stock_Units  \
0        1001       Router  Logitech  Accessories   574.15        188.0   
1        1002       Laptop        HP  Accessories  1644.89        203.0   
2        1003        Mouse  Logitech  Peripherals   642.71        325.0   
3        1004     Keyboard   Samsung  Peripherals   888.96        439.0   
4        1005     Keyboard     Canon  Electronics  1500.41         51.0   

  Country_Of_Origin Launch_Date  
0           Germany  2020-06-22  
1                UK  2023-10-15  
2             India  2019-02-22  
3               USA  2020-12-22  
4             Japan  2019-11-27

#### 2. shape : give total number of rows and columns.

In [4]:
df.shape

(515, 8)

#### 3. tail() : display last five rows.

In [5]:
df.tail()

Product_ID Product_Name   Brand     Category    Price  Stock_Units  \
510        1433      Monitor      HP  Peripherals   318.13         48.0   
511        1069      Printer  Lenovo  Peripherals   823.61        234.0   
512        1072   Smartphone    Asus  Electronics  1854.61        363.0   
513        1351     Keyboard   Apple  Accessories   801.33        340.0   
514        1123   Headphones  Lenovo  Accessories  1162.56        283.0   

    Country_Of_Origin Launch_Date  
510             India  2018-12-19  
511             China  2023-04-10  
512             India  2023-05-06  
513             India  2021-09-28  
514             Japan  2023-06-26

#### 4. info() : Summarize the dataframe.

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 515 entries, 0 to 514
Data columns (total 8 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Product_ID         515 non-null    int64  
 1   Product_Name       515 non-null    object 
 2   Brand              495 non-null    object 
 3   Category           515 non-null    object 
 4   Price              495 non-null    float64
 5   Stock_Units        494 non-null    float64
 6   Country_Of_Origin  515 non-null    object 
 7   Launch_Date        515 non-null    object 
dtypes: float64(2), int64(1), object(5)
memory usage: 32.3+ KB


#### 5. describe() : Statistically summarize the dataframe(max,min,mean,count)
#### by default provide the statistical summary for numeric column.

In [7]:
df.describe(include="all")

Product_ID Product_Name  Brand     Category        Price  \
count    515.000000          515    495          515   495.000000   
unique          NaN           10      8            3          NaN   
top             NaN       Laptop  Apple  Peripherals          NaN   
freq            NaN           63     67          182          NaN   
mean    1249.976699          NaN    NaN          NaN  1027.038485   
std      144.491687          NaN    NaN          NaN   574.943081   
min     1001.000000          NaN    NaN          NaN    52.680000   
25%     1123.500000          NaN    NaN          NaN   526.785000   
50%     1251.000000          NaN    NaN          NaN  1026.990000   
75%     1375.500000          NaN    NaN          NaN  1563.940000   
max     1500.000000          NaN    NaN          NaN  1994.630000   

        Stock_Units Country_Of_Origin Launch_Date  
count    494.000000               515         515  
unique          NaN                 6         440  
top             NaN             India  2023-04-10  
freq            NaN                94           4  
mean     256.018219               NaN         NaN  
std      144.328485               NaN         NaN  
min        0.000000               NaN         NaN  
25%      128.750000               NaN         NaN  
50%      258.000000               NaN         NaN  
75%      379.750000               NaN         NaN  
max      497.000000               NaN         NaN

##### 1.Total 8 Features and 515 Rows in this dataset.
##### 2.Product_Name have 10 unique values(laptop,mouse,keyboard,etc)
##### 3.Brand have 8 unique brands and 20 missing values.
##### 4.Category column have 3 categories.(Accessories,peripherals and electronics)
##### 5.Price column have 20 missing values.
##### 6.Stock_units have 21 missing values.
##### 7.Country have 6 different countries(India, USA,etc).
##### 8.Launch_Date have 440 dates.
##### 9.In Product laptop is most common product. It comes 63 times.
##### 10.Apple is the most common brand. (67 times)
##### 11.India launch more products than other countries(94 products).
##### 12.Average price is 1027.
##### 13.Standard deviation is less than mean. means the data values are closer to mean.
##### 14.25% of the products cost less than 526 rupees.
##### 15.50%(median) of products cost 1026 rupees.
##### 16.75% of products cost less than 1563 rupees.
##### 17.maximum price is 1994.

#### 6. columns : Returns the column name.

In [8]:
df.columns

Index(['Product_ID', 'Product_Name', 'Brand', 'Category', 'Price',
       'Stock_Units', 'Country_Of_Origin', 'Launch_Date'],
      dtype='object')

#### 7. unique(): Returns the unique values in column.

In [9]:
df['Brand'].unique()

array(['Logitech', 'HP', 'Samsung', 'Canon', 'Apple', 'Asus', 'Lenovo',
       'Dell', nan], dtype=object)

#### 8. isnull() : checks for missing values and return the output in boolean(0,1).
#### Use it with sum(), to get the total number of missing values.

In [10]:
df.isnull().sum()

Product_ID            0
Product_Name          0
Brand                20
Category              0
Price                20
Stock_Units          21
Country_Of_Origin     0
Launch_Date           0
dtype: int64

#### 9. fillna() : used to fill misssing values.

In [11]:
df['Stock_Units'].fillna(df['Stock_Units'].mean(),inplace = True)

C:\Users\Administrator\AppData\Local\Temp\ipykernel_2732\3715058198.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Stock_Units'].fillna(df['Stock_Units'].mean(),inplace = True)


In [12]:
df['Price'].fillna(df['Price'].mean(),inplace = True)

C:\Users\Administrator\AppData\Local\Temp\ipykernel_2732\790868031.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df['Price'].fillna(df['Price'].mean(),inplace = True)


In [13]:
df['Brand'] = df['Brand'].fillna('Apple')

In [14]:
df.isnull().sum()

Product_ID           0
Product_Name         0
Brand                0
Category             0
Price                0
Stock_Units          0
Country_Of_Origin    0
Launch_Date          0
dtype: int64

#### 10. sort_values : sorting the rows in ascending or descending order.

In [15]:
df.sort_values(by = 'Price', ascending = False)

Product_ID Product_Name   Brand     Category    Price  Stock_Units  \
44         1045       Laptop   Apple  Electronics  1994.63        323.0   
121        1122       Webcam      HP  Accessories  1987.56        110.0   
504        1122       Webcam      HP  Accessories  1987.56        110.0   
482        1483        Mouse   Canon  Electronics  1983.72        447.0   
184        1185       Webcam   Canon  Electronics  1979.61        386.0   
..          ...          ...     ...          ...      ...          ...   
39         1040   Headphones  Lenovo  Peripherals    63.92        263.0   
382        1383      Monitor    Asus  Peripherals    61.24        327.0   
398        1399       Laptop   Apple  Peripherals    60.53        231.0   
255        1256       Webcam    Asus  Accessories    57.36        330.0   
375        1376      Monitor    Dell  Electronics    52.68        437.0   

    Country_Of_Origin Launch_Date  
44            Germany  2023-12-30  
121           Germany  2020-09-13  
504           Germany  2020-09-13  
482               USA  2020-08-19  
184               USA  2022-04-06  
..                ...         ...  
39              China  2020-08-19  
382             China  2021-05-26  
398             India  2019-07-07  
255                UK  2021-07-17  
375             Japan  2019-12-30  

[515 rows x 8 columns]

#### 11. loc() : Access by Label.

In [16]:
df.loc[1:3, ['Product_Name','Brand','Price']]

Product_Name     Brand    Price
1       Laptop        HP  1644.89
2        Mouse  Logitech   642.71
3     Keyboard   Samsung   888.96

#### 12. iloc() : Access by position.

In [17]:
df.iloc[2]

Product_ID                  1003
Product_Name               Mouse
Brand                   Logitech
Category             Peripherals
Price                     642.71
Stock_Units                325.0
Country_Of_Origin          India
Launch_Date           2019-02-22
Name: 2, dtype: object

#### 13. groupby() : Aggregate data by category.

In [18]:
df.groupby('Brand')['Price'].mean()

Brand
Apple       1041.876172
Asus         985.467507
Canon       1016.605836
Dell        1107.049293
HP          1149.921991
Lenovo       961.197916
Logitech     996.314482
Samsung      961.732813
Name: Price, dtype: float64

#### 14. unique() : Returns the unique values in a column.

In [19]:
df['Product_Name'].unique()

array(['Router', 'Laptop', 'Mouse', 'Keyboard', 'Tablet', 'Smartphone',
       'Headphones', 'Monitor', 'Webcam', 'Printer'], dtype=object)

#### 15. nunique() : Returns the total number of unique values.

In [20]:
df.nunique()

Product_ID           500
Product_Name          10
Brand                  8
Category               3
Price                481
Stock_Units          312
Country_Of_Origin      6
Launch_Date          440
dtype: int64

#### 16. rename() :  rename the specified column or row

In [21]:
df.rename(columns={'Country_Of_Origin':'Country'}, inplace = True)

In [22]:
df.head()

Product_ID Product_Name     Brand     Category    Price  Stock_Units  \
0        1001       Router  Logitech  Accessories   574.15        188.0   
1        1002       Laptop        HP  Accessories  1644.89        203.0   
2        1003        Mouse  Logitech  Peripherals   642.71        325.0   
3        1004     Keyboard   Samsung  Peripherals   888.96        439.0   
4        1005     Keyboard     Canon  Electronics  1500.41         51.0   

   Country Launch_Date  
0  Germany  2020-06-22  
1       UK  2023-10-15  
2    India  2019-02-22  
3      USA  2020-12-22  
4    Japan  2019-11-27

#### 17. index : returns the row index.

In [23]:
df.index

RangeIndex(start=0, stop=515, step=1)

#### 18. value_count : counts unique values in a columns.

In [24]:
df['Category'].value_counts

<bound method IndexOpsMixin.value_counts of 0      Accessories
1      Accessories
2      Peripherals
3      Peripherals
4      Electronics
          ...     
510    Peripherals
511    Peripherals
512    Electronics
513    Accessories
514    Accessories
Name: Category, Length: 515, dtype: object>

#### 19. rank() : Ranks values in a column.
#### Assigning numerical positions based on rank.

In [25]:
df['Price'].rank(ascending = False)

0      382.0
1      100.0
2      361.0
3      295.0
4      146.0
       ...  
510    439.5
511    309.5
512     35.5
513    314.5
514    207.5
Name: Price, Length: 515, dtype: float64

#### 20. query() : Use sql like queries to filter data.

In [26]:
df.query("Country == 'India' and Price > 500").head()

Product_ID Product_Name     Brand     Category    Price  Stock_Units  \
2         1003        Mouse  Logitech  Peripherals   642.71   325.000000   
5         1006       Router     Apple  Electronics  1575.61   256.000000   
12        1013   Smartphone     Apple  Electronics   901.58   340.000000   
15        1016   Smartphone  Logitech  Accessories  1612.44   309.000000   
18        1019   Headphones   Samsung  Accessories  1789.77   256.018219   

   Country Launch_Date  
2    India  2019-02-22  
5    India  2021-06-13  
12   India  2018-10-27  
15   India  2021-09-30  
18   India  2022-09-06

#### 21. astype() : converts the datatype of column.

In [27]:
df["Price"].astype(float).head()

0     574.15
1    1644.89
2     642.71
3     888.96
4    1500.41
Name: Price, dtype: float64

In [28]:
len(df)

515

#### 22. filtering() :pandas allows filtering data based on filtering conditions.
####     Returns the data based on filtering conditions.

In [29]:
df[df['Price'] >= 500]

Product_ID Product_Name     Brand     Category    Price  Stock_Units  \
0          1001       Router  Logitech  Accessories   574.15        188.0   
1          1002       Laptop        HP  Accessories  1644.89        203.0   
2          1003        Mouse  Logitech  Peripherals   642.71        325.0   
3          1004     Keyboard   Samsung  Peripherals   888.96        439.0   
4          1005     Keyboard     Canon  Electronics  1500.41         51.0   
..          ...          ...       ...          ...      ...          ...   
508        1424        Mouse      Dell  Peripherals  1824.90        108.0   
511        1069      Printer    Lenovo  Peripherals   823.61        234.0   
512        1072   Smartphone      Asus  Electronics  1854.61        363.0   
513        1351     Keyboard     Apple  Accessories   801.33        340.0   
514        1123   Headphones    Lenovo  Accessories  1162.56        283.0   

     Country Launch_Date  
0    Germany  2020-06-22  
1         UK  2023-10-15  
2      India  2019-02-22  
3        USA  2020-12-22  
4      Japan  2019-11-27  
..       ...         ...  
508  Germany  2020-05-20  
511    China  2023-04-10  
512    India  2023-05-06  
513    India  2021-09-28  
514    Japan  2023-06-26  

[401 rows x 8 columns]

#### 23. pivot_table() : summarize data by using aggregate functions.(mean,median,mode,etc)

In [30]:
df.pivot_table(values = "Price", index = "Brand", columns = "Country", aggfunc = 'mean')

Country         China      Germany        India        Japan           UK  \
Brand                                                                       
Apple      916.389286  1130.992143  1094.490880  1049.843030  1104.503000   
Asus      1044.851429   807.386320  1099.368000   962.956540   883.369811   
Canon      980.450909  1108.328333   909.310000   874.098044  1112.795714   
Dell       688.563333  1203.360000  1045.471250   988.058371  1304.948333   
HP        1254.620000  1018.747135   906.389232  1089.928750  1747.873333   
Lenovo     674.731538  1004.976881  1032.900943   955.818333  1233.682727   
Logitech  1096.787500   759.175000  1024.505556  1219.898697   897.717374   
Samsung   1120.919167   846.737273   895.096154   933.965000   955.268462   

Country           USA  
Brand                  
Apple      929.421667  
Asus      1169.590909  
Canon     1164.314286  
Dell       962.913463  
HP        1266.129832  
Lenovo     898.141697  
Logitech   913.341429  
Samsung   1078.274000

In [48]:
numerical_columns = ['Price', 'Stock_Units']
print(numerical_columns)

['Price', 'Stock_Units']


In [50]:
correlation = df[numerical_columns].corr()
print(correlation)

                Price  Stock_Units
Price        1.000000    -0.005761
Stock_Units -0.005761     1.000000


#### no correlation in price and stock units.